File: apps/dsa-web/src/types/signal.ts
export interface Creator {
  id: number;
  platform: string;
  platformUid: string;
  name: string;
  category: string | null;
  isActive: boolean;
  manualWeight: number;
  fetchIntervalMin: number;
  notes: string | null;
  lastFetchAt: string | null;
  createdAt: string | null;
}

export interface CreatorCreate {
  platform?: string;
  platformUid: string;
  name: string;
  category?: string;
  isActive?: boolean;
  manualWeight?: number;
  fetchIntervalMin?: number;
  notes?: string;
}

export interface CreatorUpdate {
  name?: string;
  category?: string;
  isActive?: boolean;
  manualWeight?: number;
  fetchIntervalMin?: number;
  /** Use `null` to clear optional text fields. */
  notes?: string | null;
}

export interface QualityStats {
  totalContents: number;
  extractedCount: number;
  failedCount: number;
  pendingCount: number;
  ignoredCount: number;
  extractionSuccessRate: number;
  activeCreators: number;
  coveredCreators: number;
  creatorCoverageRate: number;
  failureExplainabilityRate: number;
}

export interface FunnelData {
  funnel: Record<string, number>;
}

export interface FailureItem {
  stage: string | null;
  reason: string | null;
  count: number;
}

export interface CreatorStats {
  creatorId: number;
  name: string;
  total: number;
  extracted: number;
  failed: number;
  successRate: number;
  lastFetchAt: string | null;
}

export interface ContentItem {
  id: number;
  creatorName: string;
  platformContentId: string;
  contentType: string;
  displayType: string;
  title: string | null;
  status: string;
  failureStage: string | null;
  failureReason: string | null;
  hasMentions: boolean;
  publishedAt: string | null;
  createdAt: string | null;
}

export interface Mention {
  id: number;
  contentId: number;
  creatorId: number;
  creatorName: string;
  assetName: string;
  assetCode: string | null;
  assetType: string;
  market: string;
  sentiment: string;
  confidence: number;
  isPrimary: boolean;
  reasoning: string | null;
  tradeAdvice: string | null;
  keyLevels: Record<string, number[]> | null;
  qualityFlags: string[];
  sourceUrl: string | null;
  publishedAt: string | null;
  createdAt: string | null;
  contentText?: string;
  transcriptText?: string;
  summaryText?: string;
}

export interface EventListParams {
  eventType?: string;
  market?: string;
  assetType?: string;
  dateFrom?: string;
  dateTo?: string;
  sortBy?: 'score' | 'created_at' | 'mention_count';
  sortOrder?: 'asc' | 'desc';
  limit?: number;
  offset?: number;
}

export interface MentionListParams {
  sentiment?: string;
  creatorId?: number;
  includeContent?: boolean;
  limit?: number;
  offset?: number;
}

export interface SignalEvent {
  id: number;
  assetName: string;
  assetCode: string | null;
  assetType: string;
  market: string;
  eventType: string;
  eventDate: string;
  score: number | null;
  bullishCount: number;
  bearishCount: number;
  neutralCount: number;
  creatorCount: number;
  mentionCount: number;
  topCreatorName: string | null;
  evidence: Record<string, unknown>[] | null;
  createdAt: string | null;
}

/** GET /signals/assets/{identifier} — overview for a single asset. */
export interface AssetDetailOverview {
  assetName: string;
  assetCode: string | null;
  assetType: string;
  market: string;
  event: {
    eventType: string;
    score: number | null;
    eventDate: string;
  } | null;
  sentimentSummary: {
    bullish: number;
    bearish: number;
    neutral: number;
  };
  creatorCount: number;
  mentionCount: number;
  creators: { id: number; name: string; weight: number }[];
}

export interface PipelineStatus {
  running: boolean;
  lastResult: Record<string, unknown> | null;
  jobId: string;
}

File: api/v1/schemas/signal.py
# -*- coding: utf-8 -*-
"""Pydantic schemas for signal API endpoints."""
from datetime import datetime, date
from typing import Optional

from pydantic import BaseModel, Field


class CreatorCreate(BaseModel):
    platform: str = "bilibili"
    platform_uid: str
    name: str
    category: Optional[str] = None
    is_active: bool = True
    manual_weight: float = Field(default=1.0, ge=0.1, le=2.0)
    fetch_interval_min: int = 60
    notes: Optional[str] = None


class CreatorUpdate(BaseModel):
    name: Optional[str] = None
    category: Optional[str] = None
    is_active: Optional[bool] = None
    manual_weight: Optional[float] = Field(default=None, ge=0.1, le=2.0)
    fetch_interval_min: Optional[int] = None
    notes: Optional[str] = None


class CreatorResponse(BaseModel):
    id: int
    platform: str
    platform_uid: str
    name: str
    category: Optional[str]
    is_active: bool
    manual_weight: float
    fetch_interval_min: int
    notes: Optional[str]
    last_fetch_at: Optional[datetime]
    created_at: Optional[datetime]

    class Config:
        from_attributes = True


class QualityStatsResponse(BaseModel):
    total_contents: int
    extracted_count: int
    failed_count: int
    pending_count: int
    ignored_count: int
    extraction_success_rate: float
    active_creators: int
    covered_creators: int
    creator_coverage_rate: float
    failure_explainability_rate: float


class FunnelResponse(BaseModel):
    funnel: dict[str, int]


class FailureItem(BaseModel):
    stage: Optional[str]
    reason: Optional[str]
    count: int


class CreatorStatsItem(BaseModel):
    creator_id: int
    name: str
    total: int
    extracted: int
    failed: int
    success_rate: float
    last_fetch_at: Optional[str]


class ContentListItem(BaseModel):
    id: int
    creator_name: str
    platform_content_id: str
    content_type: str
    display_type: str
    title: Optional[str]
    status: str
    failure_stage: Optional[str]
    failure_reason: Optional[str]
    has_mentions: bool
    published_at: Optional[datetime]
    created_at: Optional[datetime]


class MentionResponse(BaseModel):
    id: int
    content_id: int
    creator_id: int
    creator_name: str
    asset_name: str
    asset_code: Optional[str]
    asset_type: str
    market: str
    sentiment: str
    confidence: float
    is_primary: bool
    reasoning: Optional[str]
    trade_advice: Optional[str]
    key_levels: Optional[dict]
    quality_flags: list[str]
    source_url: Optional[str]
    published_at: Optional[datetime]
    created_at: Optional[datetime]
    content_text: Optional[str] = None
    transcript_text: Optional[str] = None
    summary_text: Optional[str] = None


class EventResponse(BaseModel):
    id: int
    asset_name: str
    asset_code: Optional[str]
    asset_type: str
    market: str
    event_type: str
    event_date: date
    score: Optional[float]
    bullish_count: int
    bearish_count: int
    neutral_count: int
    creator_count: int
    mention_count: int
    top_creator_name: Optional[str]
    evidence: Optional[list[dict]]
    created_at: Optional[datetime]


class PipelineRunResponse(BaseModel):
    status: str
    message: str
